tracer/src/Datadog.Trace.Manual/Baggage.cs

```diff
@@ -0,0 +1,90 @@
+﻿// <copyright file="Baggage.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System.Collections;
+using Datadog.Trace.SourceGenerators;
+using Datadog.Trace.Util;
+
+namespace Datadog.Trace;
+
+/// <summary>
+/// Baggage is a collection of name-value pairs that are propagated to downstream services.
+/// </summary>
+public static class Baggage
+{
+    private static readonly NoopDictionary NoopDictionaryInstance = new();
+
+    /// <summary>
+    /// Gets or sets the baggage collection for the current execution context.
+    /// </summary>
+    [Instrumented]
+    public static IDictionary<string, string> Current
+    {
+        get => NoopDictionaryInstance;
+        set => _ = value; // discard
+    }
+
+    private sealed class NoopDictionary : IDictionary<string, string>
+    {
+        private static readonly KeyValuePair<string, string>[] Empty = [];
+
+        int ICollection<KeyValuePair<string, string>>.Count => 0;
+
+        bool ICollection<KeyValuePair<string, string>>.IsReadOnly => true;
+
+        ICollection<string> IDictionary<string, string>.Keys => [];
+
+        ICollection<string> IDictionary<string, string>.Values => [];
+
+        string IDictionary<string, string>.this[string key]
+        {
+            get
+            {
+                ThrowHelper.ThrowKeyNotFoundException($"The key was not found: {key}");
```

Comment:
I'm torn on whether we should be doing no-ops... By doing it this way, if the user doesn't have automatic instrumentation then getting items through the accessor will _always_ throw an exception. Seems like a painful dev experience, but maybe it's better than giving them an Dictionary to mutate that's useless in the manual-only scenario?

Reply:
Yeah, good point! I must have been thinking about how how using the accessor throws when a key is not found, and since we never add any key...

> giving them an Dictionary to mutate that's useless in the manual-only scenario

Yeah, this wouldn't be horrible. If you set baggage items and then read them again later, you would expect them to be there (even if they are never propagated).

Another option could be to always return `null` or an empty string (safer?) instead of throwing an exception.

Reply:
Yeah, I missed this, we shouldn't be throwing an exception IMO. i.e. the code they run locally should "work", just return null/empty etc 

> Another option could be to always return null or an empty string (safer?) instead of throwing an exception

This is what we do elsewhere, and I think it's what we should do. If the value _could_ be `null` in production, return `null`. If it _can't_ be `null`, return `""` IMO

Reply:
Even in "production," if you lookup a key that is not in the baggage collection using the indexer instead of `TryGetValue()`, we throw a `KeyNotFoundException`, because that's the expected behavior of `IDictionary<TKey, TValue>`.
https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.idictionary-2.item

Are you're saying that can't throw an exception _at all_ even though it breaks the API contract?

Reply:
I think I'd rather return a `Dictionary<string,string>` so it behaves the same with or without automatic instrumentation. Users can set values and later get them, and looking up values that are not found will throw the expected exception in either case.

Reply:
I'm on board with returning a `Dictionary<string,string>` when there's no IL-rewriting. I'll wait for that update before approving

Reply:
What do you think of the current implementation?

Reply:
Ah should it be `Dictionary<string,string?>` ? Otherwise LGTM

Reply:
Thanks. Fixed in both the manual and internal APIs.

Reply:
> if you lookup a key that is not in the baggage collection using the indexer instead of TryGetValue(), we throw a KeyNotFoundException, because that's the expected behavior of IDictionary<TKey, TValue>.

> Are you're saying that can't throw an exception at all even though it breaks the API contract?

Oh, sorry, no, not at all. If that's the behaviour in production, it should be the behaviour without IL too